src/scheduler/mod.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright 2016 Phillip Oppermann, Calvin Lee and JJ Garzella.
+// See the README.md file at the top-level directory of this
+// distribution.
+//
+// Licensed under the MIT license <LICENSE or
+// http://opensource.org/licenses/MIT>, at your option.
+// This file may not be copied, modified, or distributed
+// except according to those terms.
+#![allow(unused)]
+
+use alloc::vec_deque::VecDeque;
+
+use sync::IrqLock;
+use interrupts::{Context,YIELD_INT};
+
+use self::thread::KThread;
+
+mod thread;
+
+// the scheduler must be created at runtime, but will not be called until
+// interrupts are enabled, which must be after the scheduler is created.
+static SCHEDULER: IrqLock<Option<Scheduler>> = IrqLock::new(None);
+
+struct Scheduler {
+    threads: VecDeque<KThread>,
+    // None => current == idle
+    current: Option<KThread>,
+    idle: KThread,
+}
+
+impl Scheduler {
+    fn new() -> Scheduler {
+        unsafe { Scheduler {
+            threads: VecDeque::new(),
+            current: Some(KThread::main()),
+            idle: KThread::idle(),
+        }}
+    }
+}
+
+pub fn init() {
+    *SCHEDULER.lock() = Some(Scheduler::new());
+}
+
+pub fn add(start: extern "C" fn()) -> Result<(), &'static str>{
+    let thread = KThread::new(start)?;
+
+    let mut lock = SCHEDULER.lock();
+    let sched = lock.as_mut().unwrap();
+    sched.threads.push_back(thread);
+    Ok(())
+}
+
+// XXX yield is a keyword
+pub fn _yield(current_stack: &'static Context) -> &'static Context {
```

Comment:
Is this only called "_yield" because of the name? Is there a more descriptive name ("process_yield" or something?) especially to distinguish this from thread_yield

Reply:
I've been wondering about what to call this. `process_yield` seems wrong because it just yields the thread. I did `_yield` because it's not really a public function like `thread_yield` and all `thread_yield` does is get `_yield` to run

Reply:
I don't really worry about name collision because they do the same thing.

Reply:
`_yield` works if it is not `pub`, but it currently is marked that way...

Reply:
It has to be called from the interrupt module (or something higher up the callstack) so it has to be pub